src/rgb21/issuer.rs

```diff
@@ -0,0 +1,203 @@
+        let mut builder = match deterministic {
+            false => ContractBuilder::with(
+                Identity::from_str(by).expect("invalid issuer identity string"),
+                Rgb21::iface(features),
+                schema,
+                main_iface_impl,
+                types,
+                scripts,
+            ),
+            true => ContractBuilder::deterministic(
+                Identity::from_str(by).expect("invalid issuer identity string"),
+                Rgb21::iface(features),
+                schema,
+                main_iface_impl,
+                types,
+                scripts,
+            ),
+        };
+
+        builder = builder
+            .add_global_state("spec", spec)
+            .expect("invalid RGB21 schema (token specification mismatch)");
+
+        Ok(Self { builder, terms })
+    }
+
+    pub fn support_liquid(mut self) -> Self {
+        self.builder = self
+            .builder
+            .add_layer1(AltLayer1::Liquid)
+            .expect("only one layer1 can be added");
+        self
+    }
+
+    pub fn add_terms(
+        mut self,
+        contract: &str,
+        media: Option<Attachment>,
+    ) -> Result<Self, InvalidRString> {
+        let terms = RicardianContract::from_str(contract)?;
+        self.terms = ContractTerms { text: terms, media };
+        Ok(self)
+    }
+
+    pub fn allocate<O: TxOutpoint>(
+        mut self,
+        method: Method,
+        beneficiary: O,
+        allocation: impl Into<Allocation>,
+    ) -> Result<Self, IssuerError> {
+        let allocation = allocation.into();
+        let beneficiary = beneficiary.map_to_xchain(|outpoint| {
+            GenesisSeal::new_random(method, outpoint.txid, outpoint.vout)
+        });
+        self.builder = self
+            .builder
+            .add_data("assetOwner", beneficiary, allocation)?;
+        Ok(self)
+    }
+
+    pub fn allocate_all<O: TxOutpoint>(
+        mut self,
+        method: Method,
+        allocations: impl IntoIterator<Item = (O, impl Into<Allocation>)>,
+    ) -> Result<Self, IssuerError> {
+        for (beneficiary, allocation) in allocations {
+            self = self.allocate(method, beneficiary, allocation)?;
+        }
+        Ok(self)
+    }
+
+    #[allow(clippy::result_large_err)]
+    pub fn issue_contract(
+        self,
+        token_index: u32,
+        image_url: &str,
+    ) -> Result<ValidContract, IssuerError> {
+        Ok(self
+            .pre_issue_contract(token_index, image_url)?
+            .issue_contract()?)
+    }
+
+    #[allow(clippy::result_large_err)]
+    fn pre_issue_contract(
+        self,
+        token_index: u32,
+        image_url: &str,
+    ) -> Result<ContractBuilder, IssuerError> {
+        let index = TokenIndex::from_inner(token_index);
+        let image_bytes = image_url.as_bytes().to_vec();
+        let preview = EmbeddedMedia {
+            ty: MediaType::with("image/*"),
```

Comment:
`image_url` is incorrect because the embedded media can be any file type. I suggest changing this method parameters to `media_bytes`, which will be an iterator of bytes, and `media_type`, which will be a `&str`.

Reply:
@cymqqqq this hasn't been completely solved, I still see an hardcoded media type (`ty: MediaType::with("image/*")`)